{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Biotite color schemes for protein sequences\n",
    "===========================================\n",
    "\n",
    "This script shows the same multiple protein sequence alignment\n",
    "in the different color schemes available in *Biotite*.\n",
    "\n",
    "    - **rainbow** - Default color scheme in *Biotite*\n",
    "    - **clustalx** - Default color scheme of the *ClustalX* software\n",
    "    - Color schemes generated with the software *Gecos* [1]_:\n",
    "        - **flower** - Light color scheme, based on *BLOSUM62*\n",
    "        - **blossom** - Light color scheme with high contrast, based on\n",
    "          *BLOSUM62*, depicts symbol similarity worse than *flower*\n",
    "        - **spring** - Light color scheme, based on *BLOSUM62*,\n",
    "          with alanine fixed to gray\n",
    "        - **wither** - Dark color scheme, analogous to *blossom*\n",
    "        - **autumn** - Dark color scheme, analogous to *spring*\n",
    "        - **sunset** - Red-green color vision deficiency adapated color\n",
    "          scheme, based on *BLOSUM62*\n",
    "        - **ocean** - Blue shifted, light color scheme,\n",
    "          based on *BLOSUM62*\n",
    "    - Color schemes adapted from *JalView* [2]_:\n",
    "        - **zappo** - Color scheme that depicts physicochemical properties\n",
    "        - **taylor** - Color scheme invented by Willie Taylor\n",
    "        - **buried** - Color scheme depicting the *buried index* \n",
    "        - **hydrophobicity** - Color scheme depicting hydrophobicity\n",
    "        - **prophelix** - Color scheme depicting secondary structure\n",
    "          propensities\n",
    "        - **propstrand** - Color scheme depicting secondary structure\n",
    "          propensities\n",
    "        - **propturn** - Color scheme depicting secondary structure\n",
    "          propensities\n",
    "\n",
    ".. [1] P Kunzmann, B Mayer,\n",
    "   \"Substitution matrix based color schemes for sequence alignment\n",
    "   visualizations.\"\n",
    "   Unpublished\n",
    "\n",
    ".. [2] M Clamp, J Cuff, SM Searle, GJ Barton,\n",
    "   \"The Jalview Java alignment editor.\"\n",
    "   Bioinformatics, 20, 426-427 (2004).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code source: Patrick Kunzmann\n",
    "# License: BSD 3 clause\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import biotite.sequence as seq\n",
    "import biotite.sequence.io.fasta as fasta\n",
    "import biotite.sequence.align as align\n",
    "import biotite.sequence.graphics as graphics\n",
    "import biotite.database.entrez as entrez\n",
    "\n",
    "\n",
    "# Generate example alignment\n",
    "# (the same as in the bacterial luciferase example)\n",
    "query =   entrez.SimpleQuery(\"luxA\", \"Gene Name\") \\\n",
    "        & entrez.SimpleQuery(\"srcdb_swiss-prot\", \"Properties\")\n",
    "uids = entrez.search(query, db_name=\"protein\")\n",
    "fasta_file = fasta.FastaFile.read(entrez.fetch_single_file(\n",
    "    uids, None, db_name=\"protein\", ret_type=\"fasta\"\n",
    "))\n",
    "sequences = [seq.ProteinSequence(seq_str) for seq_str in fasta_file.values()]\n",
    "matrix = align.SubstitutionMatrix.std_protein_matrix()\n",
    "alignment, order, _, _ = align.align_multiple(sequences, matrix)\n",
    "# Order alignment according to the guide tree\n",
    "alignment = alignment[:, order]\n",
    "alignment = alignment[220:300]\n",
    "\n",
    "# Get color scheme names\n",
    "alphabet = seq.ProteinSequence.alphabet\n",
    "schemes = [\n",
    "    \"rainbow\", \"clustalx\",\n",
    "    \"flower\", \"blossom\", \"spring\", \"wither\", \"autumn\", \"sunset\", \"ocean\",\n",
    "    \"zappo\", \"taylor\", \"buried\", \"hydrophobicity\",\n",
    "    \"prophelix\", \"propstrand\", \"propturn\"\n",
    "]\n",
    "count = len(schemes)\n",
    "# Assert that this example displays all available amino acid color schemes\n",
    "all_schemes = graphics.list_color_scheme_names(alphabet)\n",
    "assert set(schemes) == set(all_schemes)\n",
    "\n",
    "\n",
    "# Visualize each scheme using the example alignment\n",
    "fig = plt.figure(figsize=(8.0, count*2.0))\n",
    "gridspec = GridSpec(2, count)\n",
    "for i, name in enumerate(schemes):\n",
    "    for j, color_symbols in enumerate([False, True]):\n",
    "        ax = fig.add_subplot(count, 2, 2*i + j + 1)\n",
    "        if j == 0:\n",
    "            ax.set_ylabel(name)\n",
    "            alignment_part = alignment[:40]\n",
    "        else:\n",
    "            alignment_part = alignment[40:]\n",
    "        graphics.plot_alignment_type_based(\n",
    "            ax, alignment_part, symbols_per_line=len(alignment_part),\n",
    "            color_scheme=name, color_symbols=color_symbols, symbol_size=8\n",
    "        )\n",
    "fig.tight_layout()\n",
    "fig.subplots_adjust(wspace=0)\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
